wcfsetup/install/files/lib/system/style/StyleCompiler.class.php

```diff
@@ -60,38 +66,148 @@ protected function init() {
+	 * @return      string[]
+	 * @since       5.3
 	 */
-	public function compile(Style $style) {
-		$files = $this->getCoreFiles();
+	public static function getDefaultVariables() {
+		$variables = [];
+		
+		$sql = "SELECT		variable.variableName, variable.defaultValue
+			FROM		wcf".WCF_N."_style_variable variable
+			ORDER BY	variable.variableID ASC";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute();
+		$variables = $statement->fetchMap('variableName', 'defaultValue');
+		
+		// see https://github.com/WoltLab/WCF/issues/2636
+		if (empty($variables['wcfPageThemeColor'])) {
+			$variables['wcfPageThemeColor'] = $variables['wcfHeaderBackground'];
+		}
+		
+		return $variables;
+	}
+	
+	/**
+	 * Test a style with the given apiVersion, imagePath and variables. If the style is valid and does not throw an
+	 * error, true is returned. Otherwise a string with the error message.
+	 * 
+	 * @param       string          $apiVersion
+	 * @param       string          $imagePath
+	 * @param       string[]        $variables
+	 * @param       string|null     $customCustomSCSSFile
+	 * @return      bool|string
+	 * @since       5.3           
+	 */
+	public function testStyle($apiVersion, $imagePath, array $variables, $customCustomSCSSFile = null) {
+		$individualScss = '';
+		if (isset($variables['individualScss'])) {
+			$individualScss = $variables['individualScss'];
+			unset($variables['individualScss']);
+		}
+		
+		// add style image path
+		if ($imagePath) {
+			$imagePath = FileUtil::getRelativePath(WCF_DIR . 'style/', WCF_DIR . $imagePath);
+			$imagePath = FileUtil::addTrailingSlash(FileUtil::unifyDirSeparator($imagePath));
+		} 
+		else {
+			$imagePath = '../images/';
+		}
+		$variables['style_image_path'] = "'{$imagePath}'";
+		
+		// apply overrides
+		if (isset($variables['overrideScss'])) {
+			$lines = explode("\n", StringUtil::unifyNewlines($variables['overrideScss']));
+			foreach ($lines as $line) {
+				if (preg_match('~^@([a-zA-Z]+): ?([@a-zA-Z0-9 ,\.\(\)\%\#-]+);$~', $line, $matches)) {
+					$variables[$matches[1]] = $matches[2];
+				}
+			}
+			unset($variables['overrideScss']);
+		}
+		
+		// api version
+		$variables['apiVersion'] = $apiVersion;
+		
+		$parameters = ['scss' => ''];
+		EventHandler::getInstance()->fireAction($this, 'compile', $parameters);
+		
+		$files = $this->getFiles();
+		
+		if ($customCustomSCSSFile !== null) {
+			if (($customSCSSFileKey = array_search(WCF_DIR . self::FILE_GLOBAL_VALUES, $files)) !== false) {
+				unset($files[$customSCSSFileKey]);
+			}
+			
+			$files[] = $customCustomSCSSFile;
+		}
 		
-		// read stylesheets in dependency order
-		$sql = "SELECT		filename, application
+		try {
+			$this->compileStylesheetToString(
+				$files,
+				$variables,
+				$individualScss . (!empty($parameters['scss']) ? "\n" . $parameters['scss'] : ''),
+				function($content) {
+					return $content;
+				}
+			);
+		}
+		catch (\Exception $e) {
+			return $e->getMessage();
```

Comment:
I think I would be happier with returning the full Exception or `null`. It makes the return value handling less ugly, because you clearly get a truthy or falsy value (unfortunately with inverted meaning). Also getting the full Exception allows the consumer inspect the stack trace or additional values, this might be helpful for future extensions.